provider/packet/environ.go

```diff
@@ -0,0 +1,336 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package packet
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/cloudconfig/providerinit"
+	"github.com/juju/juju/core/constraints"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/environs"
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/instances"
+	envinstance "github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/provider/common"
+	"github.com/juju/juju/storage"
+	"github.com/juju/schema"
+	"github.com/juju/utils/arch"
+	"github.com/juju/version"
+	"github.com/packethost/packngo"
+	"gopkg.in/juju/environschema.v1"
+)
+
+type environConfig struct {
+	config *config.Config
+	attrs  map[string]interface{}
+}
+
+type environ struct {
+	ecfgMutex    sync.Mutex
+	ecfg         *environConfig
+	name         string
+	cloud        environscloudspec.CloudSpec
+	packetClient *packngo.Client
+}
+
+var providerInstance environProvider
+
+func (e *environ) AdoptResources(ctx context.ProviderCallContext, controllerUUID string, fromVersion version.Number) error {
+	return nil
+}
+
+func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.ProviderCallContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()}, "")
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tags map[string]string, state string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	// opt := &packngo.ListOptions{Search: tag}
+	packetTags := []string{}
```

Comment:
If you want to do filtering against a list of strings you should use the `set.Strings` (from `juju/collections/set`) which provides all set-related operations that you will need.